blackbird_python/blackbird/tests/test_listener.py

```diff
@@ -789,3 +814,42 @@ def test_parse_string(self):
         ]
 
         assert bb.operations == expected
+
+
+class TestParsingForLoops:
+    """Tests for parsing for-loops correctly"""
+
+    def test_for_loop(self, parse_input_mocked_metadata):
+        """Test that a for-loop over a list is parsed correctly"""
+        bb = parse_input_mocked_metadata(
+            "for int m in [1, 2, 3]\n\tMeasureFock() | m"
+        )
+        assert np.all(
+            bb._forvar["m"] == np.array([1, 2, 3])
+        )
+        assert bb.operations == [
+            {'op': 'MeasureFock', 'args': [], 'kwargs': {}, 'modes': [1]},
+            {'op': 'MeasureFock', 'args': [], 'kwargs': {}, 'modes': [2]},
+            {'op': 'MeasureFock', 'args': [], 'kwargs': {}, 'modes': [3]}
+        ]
+
+    def test_for_range(self, parse_input_mocked_metadata):
+        """Test that a for-loop over a range is parsed correctly"""
+        bb = parse_input_mocked_metadata(
+            "for float m in 2:10:3\n\tMZgate(m, m/2) | [0, 1]"
+        )
+        assert np.all(
+            bb._forvar["m"] == np.array([2, 5, 8])
+        )
+        assert bb.operations == [
+            {'op': 'MZgate', 'args': [2.0, 1.0], 'kwargs': {}, 'modes': [0, 1]},
+            {'op': 'MZgate', 'args': [5.0, 2.5], 'kwargs': {}, 'modes': [0, 1]},
+            {'op': 'MZgate', 'args': [8.0, 4.0], 'kwargs': {}, 'modes': [0, 1]}
+        ]
+
+    def test_wrong_type_error(self, parse_input_mocked_metadata):
+        """Test that error is raised when using mixed types in for-loop list"""
+        with pytest.raises(ValueError, match="invalid value"):
+            bb = parse_input_mocked_metadata(
+                "for int m in [1, 4.2, 9]\n\tMZgate(0, 1) | [0, 1]"
+            )
```

Comment:
Great tests! Might also be worth testing

* Indexing different shaped arrays (e.g., column vectors, matrices)

* The case where stepsize is ignored, if this is allowed (wasn't sure based on the tests above)

* Having expressions in the for loop syntax?

Reply:
Added tests for everything you mentioned. Realised that 

```python
int array vars =
    1, 2
    3, 4
```

works, while

```python
int array vars =
    1, 2
    3
```

does not; but I think this is a minor issue and can be looked into closer if/when we decide to rework the BB array type to e.g. include 2d indexes.

Reply:
I think that should actually be fine, I'm not sure

```python
int array vars =
    1, 2
    3
```
is intended to/should be allowed. Since this corresponds to a 'ragged' numpy array